records/2024-04-11.2FA-magic-link.md

```diff
@@ -0,0 +1,178 @@
+
+The fingerprint could use the user's browser attributes, IP address, HTTPS session keys, etc to form a composite key that could then be hashed to uniquely identify the user. Since the success of the attacker is time dependent, [Argon2d](https://github.com/P-H-C/phc-winner-argon2) could be a good candidate for such a hashing implementation as it uses data-depending memory access to make it resistant to GPU based brute force attacks.
+
+```mermaid
+sequenceDiagram
+    actor User
+    participant Notify
+
+    User->>+Notify: Logs in
+        create participant User's browser
+    Notify->>+User's browser: Query browser attributes
+    User's browser-->>-Notify: Serve requested attributes
+
+    Notify->>Notify: Compose attributes into fingerprint
+        create participant Argon2d
+    Notify->>+Argon2d: Request hash of the fingerprint
+    Argon2d-->>-Notify: Hashed fingerprint
+        create participant DB
+    Notify->>+DB: Store fingerprint in the database as a login event
+    DB-->>-Notify: Fingerprint stored
+    Notify->>+User's browser: Respond with request for user to check their email
+    alt User never clicks the magic link
+        loop Celery task every X interval
+            Notify->>DB: Delete expired / stale fingerprint
+        end
+    end
+    alt User clicks magic link
+        User->>User's browser: Check email & Click magic link
+        User's browser-->>-Notify: Send login request
+        Notify->>User's browser: Re-request browser attributes
+        User's browser->>Notify: Requested attributes
+        Notify->>DB: Fetch Hash
+        DB-->>Notify: Hash
+        Notify->>Argon2d: Validate fingerprint hash
+        Argon2d-->>Notify: isValid
+
+        alt Invalid hash
+            Notify->>User's browser: Login failed
+        end
+        alt Valid hash
+            Notify->>User's browser: Log user in
+            Notify->>DB: Delete login event / fingerprint
+        end
+    end
+
+```
+
+Such an approach could allow us to regain control over some scenarios that typically arise from user action and behaviour that is out of our control:
+
+* The user's device is compromised
+* The user's email inbox is compromised
+* The user forwards their 2FA link to unauthorized parties
+* The user falls victim to phishing attacks
+
+
+## Considerations
+While a unique fingerprint hashed with Argon2d could be an appropriate fallback solution to thwart these types of attacks there are several considerations that add to the scope of such an implementation:
+
+* Policy and Legal implications of collecting users' browser attributes and / or IP addresses.
```

Comment:
IP address is considered PII. See:

- https://www.canada.ca/en/transparency/privacy.html
- https://www.priv.gc.ca/en/opc-actions-and-decisions/research/explore-privacy-research/2013/ip_201305/

Collection of IP addresses needs to be justified and protected (implies additional security controls). I am not convinced that the use case here (not a core Notify functionality) can be used to justify the collection of IP addresses.